### PR TITLE
Backport sanity checks that verify that std::chrono::time agrees with time()

### PR DESCRIPTION
**Pull request by nuttycom**
Backport of the first commit of bitcoin/bitcoin#21110 and of bitcoin/bitcoin#18358

### CODE REVIEW

**Comment by nuttycom** [comment]
This builds upon #6043

**Review by daira — DISMISSED** [review]
utACK

File: src/util/time.cpp [hunk 1]
```diff
@@ -49,10 +49,10 @@ bool ChronoSanityCheck()
...
     // Check that the above zero time is actually equal to the known unix timestamp.
     struct tm epoch;
-#ifdef _MSC_VER
-    if (gmtime_s(&epoch, &time_val) != 0) {
-#else
+#ifdef HAVE_GMTIME_R
     if (gmtime_r(&time_val, &epoch) == nullptr) {
+#else
+    if (gmtime_s(&epoch, &time_val) != 0) {
 #endif
```
Comment (str4d) [inline src/util/time.cpp:56]: The upstream commit being backported was merged before `ChronoSanityCheck` was added, so why was this change not made upstream then?

Please add a `Zcash: Some description` message to the commit message that notes what the changes are relative to the upstream commit and why (so they are preserved in git history).

**Review by str4d — CHANGES_REQUESTED** [review]
Backports LGTM. Commit message needs to be altered as below, and also please add `bitcoin/bitcoin@` prefix to the `cherry picked from commit X` lines (i.e. change them to `cherry picked from commit bitcoin/bitcoin@X`) so that GitHub links them correctly.

**Pull request closed without merging**